ray-operator/controllers/ray/utils/ray_resources.go

```diff
@@ -0,0 +1,137 @@
+package utils
+
+import (
+	"strconv"
+	"strings"
+
+	rayiov1alpha1 "github.com/ray-project/kuberay/ray-operator/apis/ray/v1alpha1"
+
+	v1 "k8s.io/api/core/v1"
+	logf "sigs.k8s.io/controller-runtime/pkg/log"
+)
+
+var rayClusterLog = logf.Log.WithName("RayCluster-Controller")
+
+// Updates the user-specified rayResource spec with CPU, GPU, and memory data from the rayStartParams and the pod template.
+// CPU, GPU, and memory data from user-specified rayResourceSpec overrides data from rayStartParams.
+// CPU, GPU, and memory data from rayStartParams overrides data from the podTemplate.
+// If CPU, GPU, or memory is not present in the resource map returned by this method, Ray will attempt to
+// infer these upon "ray start".
+// The map returned by this method is used in two ways:
+// 1. It is passed as an environment variable to the Ray node. This advertises resource capacities to the Ray scheduler.
+// 2. It is written to the RayCluster CR's Status field. This allows the Ray autoscaler to access the resource capacities.
+func ComputeRayResources(
+	rayResourceSpec rayiov1alpha1.RayResources,
+	rayStartParams map[string]string,
+	rayContainerResources v1.ResourceRequirements,
+) (updatedRayResources rayiov1alpha1.RayResources) {
+
+	updatedRayResources = make(rayiov1alpha1.RayResources)
+	// Copy user CPU, GPU, memory overrides and custom resources.
+	for key, value := range rayResourceSpec {
+		updatedRayResources[key] = value
+	}
+
+	// Compute CPU unless user has already provided an override.
+	if _, ok := updatedRayResources["CPU"]; !ok {
+		// If override is not specified, look at rayStartParams and the the Ray container's resources.
+		num_cpu := computeCPU(rayStartParams, rayContainerResources)
+		if num_cpu > 0 {
+			updatedRayResources["CPU"] = num_cpu
+		}
+	}
+
+	// Compute GPU unless user has already provided an override.
+	if _, ok := updatedRayResources["GPU"]; !ok {
+		// If override is not specified, look at rayStartParams and the the Ray container's resources.
+		num_gpu := computeGPU(rayStartParams, rayContainerResources)
+		if num_gpu > 0 {
+			updatedRayResources["GPU"] = num_gpu
+		}
+	}
+
+	// Compute memory unless user has already provided an override.
+	if _, ok := updatedRayResources["memory"]; !ok {
+		// If override is not specified, look at rayStartParams and the the Ray container's resources.
+		memory := computeMemory(rayStartParams, rayContainerResources)
+		if memory > 0 {
+			updatedRayResources["memory"] = memory
+		}
+	}
+	return updatedRayResources
+}
+
+func computeCPU(rayStartParams map[string]string, rayContainerResources v1.ResourceRequirements) int64 {
+	// Try using the num-cpus rayStartParam.
+	if cpuParam, ok := rayStartParams["num-cpus"]; ok {
+		cpuInt, err := strconv.ParseInt(cpuParam, 10, 64)
```

Comment:
So one notable thing here is that Ray resources are of type `int` (or some natural number) whereas k8s resources are of type `float` (or some positive real number). 

Strictly speaking you're handling the case since that would presumable result in `err` being set, but I wonder if we should do better handling? In particular, my impression is that it's quite reasonable to have a 3.8 cpu allocation or something due to system resources and sidecars. (Should we round up? should we round down? is erroring right to avoid the ambiguity?)

I guess I just think we should be explicit (not sure exactly what the best place to document this would be) about the expected behavior here (and probably add a unit test).

Reply:
Added a comments about rounding CPU up.

Reply:
Will document the behavior in the docs as well but not in this PR.